Add function to retrieve client context for table functions

VGSML opened the pull request:
This PR adds new CAPI function to retrieve the client context for the table functions bind, it's required to get a connection Id that bind the function

VGSML committed:
feat: Add functions to retrieve client context for table functions

taniabogatsch reviewed (commented):
Looks good! Could you just add a small test to `test/api/capi/capi_table_functions.cpp`? Similar to how we test this for `duckdb_scalar_function_get_client_context` in `test/api/capi/capi_scalar_functions.cpp`.

samansmink committed:
feat: switch to using extension-ci-tools for extension builds

carlopi committed:
Wrap runner.ExecuteFile, otherwise cleanup is not properly performed

sheldonrobinson committed:
Merge branch 'duckdb:main' into main

Mytherin committed:
Re-use metadata of unaltered row groups when checkpointing a table (#18395)

Follow-up from https://github.com/duckdb/duckdb/pull/18390

This PR implements metadata re-use at the row group level, so if we are
e.g. appending to a large table we no longer rewrite the metadata of
unchanged row-groups, and instead refer to the existing metadata
on-disk. In addition, this PR also performs a few fixes where we would
eagerly load columns unnecessarily.



### Performance

Consider a database storing TPC-H SF100. The biggest table (lineitem)
has 600M rows.

```sql
CALL dbgen(sf=100);
```


Now insert a single row into the largest table (`lineitem`) and
checkpoint. We measure two times: the time of the `CHECKPOINT` command,
and the full runtime of opening the database, running the insert +
checkpoint, and closing the database.

```sql
INSERT INTO lineitem VALUES (600000001, 0, 0, 0, 0, 0, 0, 0, '', '', DATE '2000-01-01', DATE '2000-01-01', DATE '2000-01-01', '', '', '');
CHECKPOINT;
```

| Operation  | v1.3.2 |  New  |
|------------|--------|-------|
| Checkpoint | 0.5s   | 0.11s |
| Full       | 0.63s  | 0.13s |

hawkfish committed:
Merge branch 'window-sorting' into sort-aggregate

hawkfish committed:
Merge branch 'window-sorting' into naive-sorting

sheldonrobinson committed:
Tabwidth 4, 2 Tab indent

sheldonrobinson committed:
Tabwidth 4, 2 Tab indent

carlopi committed:
Merge remote-tracking branch 'origin/main' into merge_13_in_main

Mytherin committed:
Correct and consistent integer arithmetic error messages (#18393)

Clean up error messages.

- Use the same error messages in UHUGEINT and HUGEINT
- Fix operation symbols in those error messages
- Fix operation names in those error messages

Mytherin committed:
Record whether or not cross products are implicit or not, and use this for converting queries back to SQL (#18394)

For execution, whether or not a cross join is implicit or not does not
matter - but it matters for converting to SQL as implicit and explicit
cross joins are handled in different orders. This fixes an issue where
`ToString()` would emit a query that was not re-executable.

Mytherin committed:
CI: Fix Discussion mirroring (#18397)

This PR fixes discussion mirroring.

The `gh` CLI tool has limited support to work with discussions
(https://github.com/cli/cli/discussions/4212), so the recommendation is
to script using GraphQL. However, we actually do not need to do much
with the discussions to mirror them: we simply need to open an internal
issue. There is no need to adjust the labels of discussions.

Mytherin committed:
Store extra metadata blocks in RowGroupPointer, and only flush dirty Metadata blocks (#18398)

Follow-up from https://github.com/duckdb/duckdb/pull/18395

This PR adds storage for `extra_metadata_blocks` in the row-group
metadata when the latest storage version (v1.4.0) is used. This is a
list of metadata blocks that are referenced by the row group metadata,
but **not** present in the list of data pointers (`data_pointers`).
These blocks can be present when either (1) columns are very wide due to
e.g. being deeply nested, or (2) the final column pointer "crosses" the
metadata block threshold, as `data_pointers` points only to the
beginning of the column metadata.

Usually this list is empty - and therefore storing this does not take up
a lot of extra storage space. The presence of this list allows us to
more efficiently re-use metadata as we know exactly which metadata
blocks a row group points to, without having to do any additional
deserialization.

### Only Flush Dirty Metadata blocks

Previously our metadata manager would flush all metadata blocks,
incurring a lot of unnecessary I/O now that we are doing a lot of
metadata re-use. This PR reworks this so that we keep track of which
blocks are dirty and only flush the dirty blocks.

### Performance

Running the benchmark in https://github.com/duckdb/duckdb/pull/18395
again, we now get the following timings:

| Operation  | v1.3.2 |  Re-Use  |  New  |
|------------|--------|-------|-------|
| Checkpoint | 0.5s   | 0.11s | 0.04s |
| Full       | 0.63s  | 0.13s | 0.07s |

ZuleykhaPavlichenkova-TomTom committed:
ff

Mytherin committed:
Use GetSetting/GetEnum for more settings

Maxxen committed:
error on unresolved template types after the bind callback

taniabogatsch committed:
Merge branch 'main' into index-append-buffer

Maxxen committed:
add overload for list_value

Maxxen committed:
fix error msg

taniabogatsch committed:
move index binding into bind_insert

Mytherin committed:
Internal #3273: Window Hashed Sort (#18337)

* implement materialized sorting interface
* add materialization to cdc
* Create hashed sort infrastructure
* Add Finalize call
* Add merged callback (for linear mask computation)
* First passing test!
* Create hash groups during Combine
* Remove redundant (and incorrect) counting
* Correctly scan sort columns that are not at the front of the payload
* Materialise sort expressions
* Handle no distinct values when computing masks
* Move OVER() hash group construction to Finalize
* Optimize single sort keys
* Correctly call DistinctFrom for ORDER BY keys
* Finish conversion of the collection scanner to pure chunks.
* Explicitly ReferenceStructColumns in ComputeMasks instead of trying to
guess.
* Apply BatchedDataCollection patch.
* Disable unrelated GetAllNeighborSets asserts.
* Fix allocation threading in WindowDistinctAggregatorLocalState
* Fix non-deterministic window tests with unstable sorting
* Fix amalgamation build header problem.

Mytherin committed:
Allow extension settings to have a defined default scope (local or global)

Mytherin committed:
Remove incorrect assertion

Mytherin committed:
Add DuckLake back in

szarnyasg committed:
CI: Add separate job for discussion mirroring

Maxxen committed:
improve error messages

samansmink committed:
switch to main extension-ci-tools

pdet committed:
When resizing we only need to care if it's same sign and MSB of MSD is set

pdet committed:
Dont trim 0's

samansmink committed:
small fixes

samansmink committed:
try solve boolean bs

samansmink committed:
remove commented out needs dep

Mytherin committed:
Fix extension entries

Mytherin committed:
Wrap runner.ExecuteFile, otherwise cleanup is not properly performed (#18400)

Problem is that when using a non-standard folder and there is a throw
that escape, cleanup is not properly performed and cause subsequent
tests to be broken.

Mytherin committed:
[BUGFIX] Update delim offset for RHS of DELIM JOIN when correlated column is in RHS of Cross product (#18375)

Fixes: https://github.com/duckdblabs/duckdb-internal/issues/5331
Fixes: https://github.com/duckdb/duckdb/issues/18267

Also fixes https://github.com/duckdb/duckdb/issues/17701 (Instead of an
Internal exception it throws a Conversion Error)

This is similar to how we set the delim offset for inner joins as well.

taniabogatsch committed:
binding phase atomic

samansmink committed:
fix ci failures

Mytherin committed:
CI: Add separate job for discussion mirroring (#18407)

Followup of #18397, splitting this to two separate jobs

Mytherin committed:
[ Python SQLLogic Tester ] Add `MERGE_INTO` statement to duckdb python (#18402)

This PR adds support of new `MERGE_INTO` statement by Python SQLLogic
Tester, so it doesn't fail with `INTERNAL Error: Unrecognized
StatementType in ExpectedResultType: MERGE_INTO`

Should fix https://github.com/duckdblabs/duckdb-internal/issues/5371

pdet committed:
wip

Mytherin committed:
Set parameter correctly

samansmink committed:
cidebug: checkout to PR merge commit through ci tools

samansmink committed:
remove accidental ;

Maxxen committed:
throw on unsupported inference cases

Maxxen committed:
Merge branch 'main' of https://github.com/duckdb/duckdb into template-types

taniabogatsch committed:
nits

samansmink committed:
cidebug: retrigger

Mytherin committed:
Remove incorrect assertion (#18404)

Mytherin committed:
Turn ArrowFormatVersion into an enum class

samansmink committed:
cidebug: retrigger

Mytherin committed:
Move the variant_legacy_encoding setting to the Parquet reader

pdet committed:
Sasageyo

hawkfish committed:
Merge branch 'main' into sort-aggregate

Mytherin committed:
Internal #5294: TIME_NS C API (#18215)

* Add enum, struct and basic value APIs

fixes: duckdblabs/duckdb-internal#5294

pdet committed:
Woo all tests are currently passing

samansmink committed:
restore citools repo and branch

hawkfish committed:
Merge branch 'main' into naive-sorting

Maxxen committed:
adjust tests

Maxxen committed:
fix map_extract behavior to always return null if input is null

pdet committed:
Some cleanup

pdet committed:
More tests

samansmink committed:
merge out-of-tree extensions with in-tree extensions build

pdet committed:
WIP

Maxxen committed:
tidy

Maxxen committed:
Merge branch 'main' of https://github.com/duckdb/duckdb into template-types

Commit:
Internal #5368: WindowNaiveAggregator Sort Update

* PR Feedback (comment nit).

Commit:
Internal #5384: WindowDistinctAggregator Sort Update

* Refactor ColumnDataCollectionScanner as WindowCollectionChunkScanner
* Replace sequential iterator scan with vectorised implementation.
* Cut holes for Sort APIs.
* Untested!

Mytherin committed:
Add DuckLake back in (#18405)

rustyconover committed:
fix: remove anything following ? in database name

To support databases that are specified using URL paths with
query parameters, I'd like to propose that when starting duckdb
with

`airport:grpc://localhost:50003/test1?option=abc`

The database name is just `test1` rather than `test1?option=abc`.

hawkfish committed:
Internal #5384: WindowDistinctAggregator Sort Update

* Fix crashes

Mytherin committed:
Add support for table_constraints of AdbcConnectionGetObjects() (#18181)

See

https://arrow.apache.org/adbc/19/cpp/api/group__adbc-connection-metadata.html#ga32f64f4cfc86140e2d8d02b328215dbf
for `AdbcConnectionGetObjects()` and `table_constraints` specification.

pdet committed:
Make realloc rule tighter for lineitem

Mytherin committed:
Correctly fetch only base column data in ColumnData::FetchUpdateData

taniabogatsch committed:
add concurrent test and patch spatial

taniabogatsch committed:
Merge branch 'main' into index-append-buffer

pdet committed:
Add benchmark

Mytherin committed:
Merge `v1.3-ossivalis` in `main` (#18401)

Trivial after https://github.com/duckdb/duckdb/pull/18388.

Most extensions still outdated, but not httpfs.

samansmink committed:
nit: remove submodules=true

pdet committed:
Fix build

samansmink committed:
fix comment

Mytherin committed:
feat: remove anything following `?` in database name (#18417)

To support databases that are specified using URL paths with query
parameters, I'd like to propose that when starting duckdb with

`airport:grpc://localhost:50003/test1?option=abc`

The database name is just `test1` rather than `test1?option=abc`.

taniabogatsch committed:
adjust block size tests

taniabogatsch committed:
Merge branch 'main' into block-size-fixes

taniabogatsch committed:
Merge branch 'main' into index-append-buffer

pdet committed:
only siths deal in absolutes

taniabogatsch committed:
tidy test configs

Mytherin committed:
Correctly fetch only base column data in ColumnData::FetchUpdateData (#18423)

Prevents reading the validity data twice

This doesn't actually cause incorrect behavior - but it triggers an
assertion, and is unnecessary work.

pdet committed:
Gotta flip these bits

taniabogatsch committed:
add reason to skipped tests

Mytherin committed:
Refactor extension CI to use extension-ci-tools (#18361)

In this PR, we switch the extension builds in CI to fully use the
[extension-ci-tools](https://github.com/duckdb/extension-ci-tools) repo.

This has some benefits:
- removes the CI code duplication betwee duckdb and extension-ci-tools.
This means that every extension (both in-tree and out-of-tree) is now
built using the
[extension_distribution](https://github.com/duckdb/extension-ci-tools/blob/main/.github/workflows/_extension_distribution.yml)
workflow.
- simplifies extension builds into single workflow
- produces a single zip files containing the extensions that can be
easily deployed and used in further testing
- easily allows building the delta extension in duckdb CI again (was
disabled before due to running out of disk space)

This PR also modifies the behaviour for running Extension CI in PRs:
- for OSX we only run `osx_arm64`
- for linux we only run `linux_amd64`
- for wasm only run `wasm_mvp`
- for windows we only run `windows_amd64`

taniabogatsch committed:
adjust tests

hawkfish committed:
Internal #5384: WindowDistinctAggregator Sort Update

* Remove unnecessary range patching

pdet committed:
Casts + run all sum tests with varints

Commit:
Internal #5384: WindowDistinctAggregator Sort Update

* Don't assume combine operations will use the local state's local sort sink.

Commit:
Merge branch 'main' into naive-sorting

Commit:
Merge branch 'naive-sorting' into window-distinct

Mytherin committed:
Internal #5367: SortedAggregateFunction Sort Update  (#18408)

* Convert SortedAggregateFunction to the new sort code.
* Fix sign warning

Mytherin committed:
Internal #5368: WindowNaiveAggregator Sort Update  (#18409)

* Convert sort logic for ORDER BY arguments
* Plumb ExecutionContext through window local states.
* Plumb ExecutionContext through window local APIs instead of holding
onto it from the constructor.

Mytherin committed:
[Fix] Block size nightly (#18425)

Nightly run here:
https://github.com/taniabogatsch/duckdb/actions/runs/16568441774

Related issue: https://github.com/duckdblabs/duckdb-internal/issues/5378

Mytherin committed:
[Chore] Tidy test configs (#18426)

While looking into
https://github.com/duckdblabs/duckdb-internal/issues/5207, I noticed
that the descriptions and test config files had already diverged a bit
and decided to unify them.

mapleFU committed:
Parquet: add row-group ordinal during writing

mapleFU committed:
Resolve comments

maiadegraaf committed:
Add QueryContext to FileHandle::Read(..., idx_t location)

Mytherin committed:
Initial implementation of generic settings, migrate default_order and default_null_order to generic settings

Mytherin committed:
Rework SET/RESET test to use SQL statements instead of option internals

Mytherin committed:
Several test fixes + fix for lookups

Mytherin committed:
Handle aliases separately

Mytherin committed:
Separate aliases from options, and display these separately in duckdb_settings()

Mytherin committed:
Test fixes

Mytherin committed:
Make SetOption work with new generic options

Mytherin committed:
Increase size threshold in test

Mytherin committed:
Move more settings to new generic settings

taniabogatsch committed:
improve skipped tests and add verify_fetch_row config

taniabogatsch committed:
Merge branch 'main' into index-append-buffer

taniabogatsch committed:
patch vss

pdet committed:
Lots of rewrites to use VarintIntermediate struct with simpler metadata

Mytherin committed:
Clean up includes and fix tests

Mytherin committed:
Add missing includes

Mytherin committed:
Move allow_extensions_metadata_mismatch to a generic setting

maiadegraaf committed:
clean up

Mytherin committed:
Move arrow settings to be generic settings

pdet committed:
Small fixes

Mytherin committed:
More generic settings, clean up enum accessors

Maxxen committed:
Merge branch 'main' of https://github.com/duckdb/duckdb into template-types

Maxxen committed:
Update src/function/function_binder.cpp

Co-authored-by: Gabor Szarnyas <szarnyasg@gmail.com>

Maxxen committed:
Merge branch 'template-types' of https://github.com/Maxxen/duckdb into template-types

pdet committed:
Couple opts

Maxxen committed:
switch template to take explicit names

taniabogatsch committed:
fix create index bug

Maxxen committed:
fixed tanias comments

pdet committed:
remove old comments

Mytherin committed:
Include pyodide build configuration (#18183)

This adds build support for the latest ABI version of pyodide.

I wasn't sure how to get `duckdb/duckdb-pyodide` to build for the latest
ABI version of pyodide / emcripten so I figured I'd make this PR to
bring it to your attention:
https://github.com/duckdb/duckdb-pyodide/issues/7

Mytherin committed:
set_variables

Mytherin committed:
Move more settings to generic settings

Mytherin committed:
Move more settings to generic

Mytherin committed:
Parquet: add row-group ordinal during writing encryption (#18433)

RowGroup.ordinal is required in encryption, see[1][2]. Arrow-cpp[3] and
arrow-rs writes this

[1] https://github.com/apache/parquet-format/pull/453
[2] https://github.com/apache/parquet-format/blob/master/Encryption.md
[3]
https://github.com/apache/arrow/commit/f4a63d41ebbc57566f215c1d1e87fc1647071dae
[4]
https://github.com/apache/arrow-rs/blob/1d9afbc037d7c0562b7f80115928a1b5050c5692/parquet/src/file/writer.rs#L736

Mytherin committed:
Add missing includes

Mytherin committed:
_t

pdet committed:
Add function that allocates only at result to ADD()

Mytherin committed:
Amalgamation fix

pdet committed:
Include-fest

pdet committed:
tidy

pdet committed:
Merge remote-tracking branch 'upstream/main' into varint_physical

rgbkrk committed:
Update pyodide build to 0.28.0

sheldonrobinson committed:
Merge branch 'duckdb:main' into main

Mytherin committed:
Test + include fix

Mytherin committed:
[Fix] Reset segment memory when initialising new Prefix (#18441)

When making changes to indexes, we might reuse previously used index
segments. In the case of this bug, we reuse a previous segment in a
place where we then expect it to be uninitialised.

I am not sure how to add the reproducer, since the table (with the
correct sequence of values to cause this) is 10MB (compressed, all other
columns removed).

Fixes https://github.com/duckdb/duckdb/issues/18190

Mytherin committed:
Format

Mytherin committed:
Update pyodide build to 0.28.0 (#18446)

We can use the latest release of pyodide for builds now.

Mytherin committed:
Move OnConflictClause data to separate struct

Mytherin committed:
skip_reload

Mytherin committed:
Extension entries

Mytherin committed:
Add support for "template" types (#18410)

This PR adds basic support for defining functions with "template" types.

Template types function like "placeholder" types when defining a
function, similarly to the already existing `ANY` type. The main
difference is that instead of relying on the function to implement a
custom "bind" callback in order to replace the `ANY` type with a
concrete type after inspecting the function arguments, `TEMPLATE` types
are automatically "inferred" to the most general type of all types
passed to them, and replaced by the binder, before the function
callbacks are invoked. Additionally, it's possible to declare a function
with multiple _different_ `TEMPLATE` types by declaring them with an
"id".

## Consistent inference logic and errors

In particular, functions that operate on nested types have previously
required complex binding logic, and generally struggle with handling
`SQLNULL` and `UNKNOWN` (from prepared statements) properly. By using
template types, all logic to deal with inferring argument/return types
is now consistent and centralized in one location.

The actual inference logic is pretty simple and consist of two steps,
__Inference__ and __Substitution__.

In the __Inference__ phase, we recursively visit the type in a function
parameter simultaneously as the type of the corresponding argument, and
compare any nested types "positionally'. Once we find a template (child)
type on the parameters side, we "bind" the template "id" to the
corresponding (child) type on the argument side. If the template was
already bound, and the bound type differs from the type on the argument
side, we try to "promote" the bound type with the new type using
`MaxLogicalType` to "widen" the inferred template type. If the promotion
fails, we throw an error explaining the inference steps so far, which is
now consistent across functions and much more detailed, for example:

```sql
select list_value(1, 1, 1, '1', 1.10, [2]);
----
Binder Error:
Cannot deduce template type 'T1' in function: 'list_value(T1, [T1...]) -> T1[]'
Type 'T1' was inferred to be:
 - 'INTEGER_LITERAL', from first occurence
 - 'INTEGER', by promoting 'INTEGER_LITERAL' + 'STRING_LITERAL'
 - 'DECIMAL(12,2)', by promoting 'INTEGER' + 'DECIMAL(3,2)'
 - 'INTEGER[]', which is incompatible with previously inferred type!

LINE 1: select list_value(1, 1, 1, '1', 1.10, [2]);
                                               ^
```

Once we've tried to bind all template types, we call the functions
"bind" callback (if present) to give the callback a chance to resolve
any unbound templated, if they would like to.

Then, in the __Substitution__ phase, we again iterate through and visit
all function parameter types, but also the function "varargs" type, and
return type, and now actually replace the template types with their
inferred types. If we find a template which does not have a type
inferred for it, we also throw an error.

### Better function signatures

The printed signatures of generic functions are much more informative,
as it's now possible to actually distinguish the relationship between
different generic (templated) parameters, and sometimes even possible to
identify the behavior of certain edge cases (e.g. `list_value()`).
Here's some examples:

- `map_contains(ANY, ANY) -> ANY` => `map_contains(MAP(T1, T2), T1) ->
T1`
- `list_value(ANY) -> LIST` => `list_value(T1, [T1...]) -> T1[]` and
`list_value() -> NULL[]`

### Behavior changes

By replacing the custom binding logic in a lot of functions with
template types instead, I've found that some of the old behavior was
inconsistent in a way that I would consider buggy. I've thus made some
adjustments.

- There is now a separate overland for the "empty" `MAP()` case, instead
of trying to change the behavior in the bind by inspecting if there are
0 or 2 arguments.
- `list_value()` now has a separate overload for the "empty list" case
when no arguments are passed.
- Lots of map functions .e.g `map_keys`, `map_values`, `map_extract`
used to be defined as taking varargs, but then check and error in the
bind callback if there are more than the expected amount of arguments.
This has now changed so that they are no longer varargs and have well
defined signatures
- `map_extract` used to return `[]` when passed `SQLNULL`, but `NULL`
when passed any typed null. This is has now changed to always return
null regardless if passed a typed or untyped null.
- `map_extract` used to accept anything as the key lookup argument, even
if the type didn't match the actual type of the key, and then return
`NULL`. This now throws a proper type error.
- `array_agg()`/`list()` used to set their return type to `SQLNULL` when
passed a prepared statement parameter. Since they are now defined with
template types they now properly throw `ParameterNotResolved`
exceptions.

### Limitations

Only templated child types in "unnamed" `STRUCT` can be inferred for
now. `UNION` types are also not supported yet. This doesn't mean you
can't pass a `STRUCT` or `UNION` to a template - you just can't declare
a function with a signature like `foo(STRUCT(x T1, i INT), arg INT) ->
T1`. DuckDB does not have any of these in core, so it's a bit difficult
to test, but we want to enable this so extensions can define functions
like this in a follow-up PR.

I also haven't changed all of our nested functions that use `ANY` to use
`TEMPLATE` yet, as some are very advanced and I don't want to risk
breaking them yet.

### Future work

There's lots more I want to do with this in the future:

- Support templated __non-type type-parameters__ as well, so that you
can define e.g .`array_concat(INT[N1], INT[N2]) -> INT[N1 + N2]` or
`decimal_add(DECIMAL(W1, S1), DECIMAL(W1, S1)) -> DECIMAL(W1, S1)`

- Support for typing __lambdas__, e.g. `list_select(T1[], (T1 ->
BOOLEAN)) -> T1[]`

- Support for __type constraints/traits__, e.g. `list_sort(T1[]) where
T1: Sortable` or `safe_add(T1, T1) -> T1 where T1: Numeric`

Mytherin committed:
Merge branch 'main' into settingsrework

Mytherin committed:
Internal #5384: WindowDistinctAggregator Sort Update  (#18442)

* Refactor ColumnDataCollectionScanner as WindowCollectionChunkScanner
* Replace sequential iterator scan with vectorised implementation.
* Cut holes for Sort APIs.
* Remove unnecessary range patching
* Don't assume combine operations will use the local state's local sort
sink.

Mytherin committed:
Add log manager include back

Mytherin committed:
[Chore] Improve skipped tests in test config and add verify_fetch_row config (#18436)

New `verify_fetch_row.json` config:
```json
  "description": "Run on persistent databases as storage with row verification enabled.",
  "initial_db": "{TEST_DIR}/{BASE_TEST_NAME}__test__config__verify_fetch_row.db",
  "on_init": "PRAGMA verify_fetch_row;",
  "skip_compiled": "true",
  "skip_tests": [
```

Close https://github.com/duckdblabs/duckdb-internal/issues/5207.

New structure of skipped tests:
```
LogicalType::LIST(
         LogicalType::STRUCT(
{{"reason", LogicalType::VARCHAR}, 
{"paths", LogicalType::LIST(LogicalType::VARCHAR)}})),
```

Example:
```json
    {
      "reason": "Contains FIRST (non-deterministic).",
      "paths": [
        "test/sql/parallelism/intraquery/test_parallel_nested_aggregates.test"
      ]
    },
    {
      "reason": "FIXME: Unexpected parsed statement verifier failure.",
      "paths": [
        "test/sql/types/decimal/large_decimal_constants.test",
        "test/sql/types/uhugeint/test_uhugeint_conversion.test",
        "test/sql/types/hugeint/test_hugeint_conversion.test"
      ]
    },
```

There are a bunch of `FIXME` in the new `verify_fetch_row.json` - I'll
open the respective internal issues once we've merged this PR.

Mytherin committed:
Buffer index appends during WAL replay (#18313)

Prior to this PR, during WAL replay, if we pushed appends to the WAL,
then we would try to bind the index to append these. Now, we buffer them
in the `UnboundIndex`, which prevents any index binding steps during WAL
replay.

Close https://github.com/duckdblabs/duckdb-internal/issues/3772

Mytherin committed:
WIP: initial transform of ON CONFLICT into MERGE statement working

taniabogatsch committed:
Merge branch 'v1.3-ossivalis' into merge-ossivalis

Mytherin committed:
Add support for generic settings, and move many settings over to generic settings (#18447)


In the current settings configuration, settings are hard-coded options
in the `DBConfig` or `ClientConfig` respectively.
As a result, settings can either be *local* or *global* - they cannot be
set in both scopes.
In addition, for every setting, we need to generate code in order to
interact with the hard-coded option.

This PR adds support for generic settings to the settings generation
script. These settings operate similar to extension settings:

* They are stored in the `set_variables` list when set
* They can be set both locally and globally

#### Generic Settings

Here is an example of a generic setting:

```json
{
    "name": "pivot_limit",
    "description": "The maximum number of pivot columns in a pivot statement",
    "type": "UBIGINT",
    "default_scope": "local",
    "default_value": "100000"
}
```

The two keywords are `default_scope` and `default_value`.
* `default_value` is the value returned when the setting is not set.
* `default_scope` is the scope used when none is explicitly specified by
the user (i.e. when running `SET pivot_limit=...`)


The setting can be read using the generated struct `PivotLimitSetting`:

```cpp
auto pivot_limit = DBConfig::GetSetting<PivotLimitSetting>(context);
```

This will return the type as defined in the JSON (in this case `idx_t`).
It will automatically do the "correct" cascading - i.e. first reading
the connection-local setting, followed by reading the global setting,
followed by getting the default value.

#### Enums

Enums are also supported using the `ENUM<ENUM_TYPE>` type, e.g.:

```json
{
    "name": "debug_verify_vector",
    "description": "DEBUG SETTING: enable vector verification",
    "type": "ENUM<DebugVectorVerification>",
    "default_scope": "global",
    "default_value": "NONE"
}
```

The `GetSetting` will return the underlying enum. The setting is stored
as a `VARCHAR`. An `OnSet` method is automatically generated that
verifies only correct enum values are passed in:

```cpp
void DebugVerifyVectorSetting::OnSet(SettingCallbackInfo &info, Value &parameter) {
	EnumUtil::FromString<DebugVectorVerification>(StringValue::Get(parameter));
}
```


#### Aliases

This PR also reworks the way that setting aliases work. Previously
setting aliases would just duplicate the setting (with both pointing to
the same underlying setting). We now generate explicit aliases. The list
of aliases is also returned in `duckdb_settings`, e.g.:

```sql
select * from duckdb_settings() where name='default_null_order';
┌────────────────────┬────────────┬───────────────────────────────────────────────────────────────────────┬────────────┬─────────┬──────────────┐
│        name        │   value    │                              description                              │ input_type │  scope  │   aliases    │
│      varchar       │  varchar   │                                varchar                                │  varchar   │ varchar │  varchar[]   │
├────────────────────┼────────────┼───────────────────────────────────────────────────────────────────────┼────────────┼─────────┼──────────────┤
│ default_null_order │ NULLS_LAST │ NULL ordering used when none is specified (NULLS_FIRST or NULLS_LAST) │ VARCHAR    │ GLOBAL  │ [null_order] │
└────────────────────┴────────────┴───────────────────────────────────────────────────────────────────────┴────────────┴─────────┴──────────────┘
```

Previously, aliases would create a duplicate of the setting, e.g.:

```sql
select * from duckdb_settings() where name like '%null_order';
┌────────────────────┬────────────┬───────────────────────────────────────────────────────────────────────┬────────────┬─────────┐
│        name        │   value    │                              description                              │ input_type │  scope  │
│      varchar       │  varchar   │                                varchar                                │  varchar   │ varchar │
├────────────────────┼────────────┼───────────────────────────────────────────────────────────────────────┼────────────┼─────────┤
│ default_null_order │ nulls_last │ NULL ordering used when none is specified (NULLS_FIRST or NULLS_LAST) │ VARCHAR    │ GLOBAL  │
│ null_order         │ nulls_last │ NULL ordering used when none is specified (NULLS_FIRST or NULLS_LAST) │ VARCHAR    │ GLOBAL  │
└────────────────────┴────────────┴───────────────────────────────────────────────────────────────────────┴────────────┴─────────┘
```

#### Future Work

This PR moves many options to be generic settings - but the old setting
methods are still kept around.
Ideally we move all options to the new format in the future.
In order to facilitate that I think we will also want to add the ability
to force options to only have a certain scope, e.g. for some options
like `threads` or `memory_limit` it does not make sense to have
connection-only variants.

As this PR is currently large enough I leave that for a future
iteration.

Mytherin committed:
Fixes for lambdas in upsert

Mytherin committed:
INSERT OR REPLACE working

pdet committed:
PR WIP

Commit:
Internal #5385: WindowMergeSortTree Sort Update

* Convert WindowMergeSortTree, WindowIndexTree and WindowTokenTree to the new sort code
* Cut more holes for window local states to have access to the ExecutionContext

fixes: duckdblabs/duckdb-internal#5385
fixes: duckdblabs/duckdb-internal#5386

Commit:
Internal #5385: WindowMergeSortTree Sort Update

* Update enum serialisation

fixes: duckdblabs/duckdb-internal#5385
fixes: duckdblabs/duckdb-internal#5386

Commit:
Internal #5366: WindowDeltaScanner

* Refactor WindowTokenTreeLocalState::BuildLeaves into a template

Commit:
Internal #5366: WindowDeltaScanner

* Refactor WindowHashGroup::ComputeMasks into a template

Commit:
Internal #5366: WindowDeltaScanner

* Refactor WindowDistinctAggregatorLocalState::Sorted into a template

Mytherin committed:
Internal #5385: WindowMergeSortTree Sort Update (#18461)

* Convert WindowMergeSortTree, WindowIndexTree and WindowTokenTree to
the new sort code
* Cut more holes for window local states to have access to the
ExecutionContext
* Update enum serialisation

fixes: duckdblabs/duckdb-internal#5385
fixes: duckdblabs/duckdb-internal#5386

Mytherin committed:
Bump postgres to latest main

Mytherin committed:
Several test fixes after WHERE clause fix

taniabogatsch committed:
Merge branch 'main' into merge-ossivalis

Mytherin committed:
Fix error message in projection binder

Mytherin committed:
Generate single AND/OR chains

taniabogatsch committed:
fix test for configs

Mytherin committed:
Expand default values in values list

Mytherin committed:
Add extension entry

Mytherin committed:
Add DISTINCT ON (unique_cols), and fix for UPDATE with generated columns

Mytherin committed:
Test fixes + add update/generated column test

pdet committed:
varint_t everything that is varint_t

pdet committed:
Adding value_unsafe<varint_t>

hawkfish committed:
Merge branch 'main' into sorting-delta

maiadegraaf committed:
Refactor QueryContext usage in Catalog and FileHandle methods

Mytherin committed:
Bump postgres to latest main (#18464)

Mytherin committed:
Merge ossivalis (#18456)

maiadegraaf committed:
Add default OpenFile() without QueryContext

Mytherin committed:
Internal #5366: WindowDeltaScanner  (#18468)

* Refactor WindowTokenTreeLocalState::BuildLeaves into a template
* Refactor WindowHashGroup::ComputeMasks into a template
* Refactor WindowDistinctAggregatorLocalState::Sorted into a template

Mytherin committed:
Add missing includes

taniabogatsch committed:
don't cancel rollback of other transaction upon exception

taniabogatsch committed:
use merge

taniabogatsch committed:
more insights around dict fsst compression failure

Mytherin committed:
Patch ducklake

taniabogatsch committed:
rename

Mytherin committed:
Add benchmarks

Mytherin committed:
More patches

Mytherin committed:
SUM and + Operator for Varints (#18424)

This PR adds the `+` and SUM operators for VARINTs.

The addition is implemented in-place, and when necessary, the target
VARINT will be reallocated to twice the required space to avoid future
reallocations. It is important to note that having more bytes than
necessary actually breaks our VARINT definition. So, before adding the
VARINT to a vector, we must `TRIM()` it to restore a valid format.

This implementation can still be heavily optimized:
1. The 2x reallocation might be overly aggressive.
2. We currently check absolute values to determine the sign, which could
be improved with type pruning. The same applies to `GetStartDataPos`.
3. We don't necessarily need to reallocate during `TRIM`. While this
should only happen once during `SUM`, it might impact the `+` operator
more if used in a chain.

I've compared the SUM implementation in
benchmark/types/varint/varint_sum.benchmark with:

```python
big_number_str = "1" + "0" * 10000000
big_number = int(big_number_str)

n = 10_000
big_numbers = [big_number] * n

durations = []
for _ in range(5):
    start = time.time()
    total_sum = sum(big_numbers)
    end = time.time()
    durations.append(end - start)

total_sum_str = str(total_sum)
median_time = statistics.median(durations)

print(median_time)
```

The Python implementation runs in ~0.15 seconds, while ours runs in ~1.1
seconds. So, there is definitely still room for improvement.

samansmink committed:
re-nable extensions in invokeci

samansmink committed:
fix: replace all wasm references

samansmink committed:
fix: remove reference to removed R workflow too

samansmink committed:
fix: remove ALL references to the R workflow

taniabogatsch committed:
format fix

Mytherin committed:
Merge branch 'main' into upsertmerge

Mytherin committed:
Limit polars version

Mytherin committed:
[Fix] Rework transaction logic in commit, rollback and checkpoint paths (#18474)

Close https://github.com/duckdblabs/duckdb-internal/issues/5437

hawkfish committed:
Internal #5384: Window Sorting Polish

* Ensure that sorting tasks are marked as completed when execptions are thrown
* Pass ExecutionContext in to HashedSort APIs.

Mytherin committed:
Fix test

Mytherin committed:
re-nable extensions in invokeci (#18476)

This was accidentally borked with
https://github.com/duckdb/duckdb/pull/18361

Mytherin committed:
Internal #5384: Window Sorting Polish (#18484)

* Ensure that sorting tasks are marked as completed when exceptions are
thrown
* Pass ExecutionContext in to HashedSort APIs.

sheldonrobinson committed:
Merge branch 'duckdb:main' into main

Mytherin committed:
Unify `ON CONFLICT` and `MERGE INTO` (#18480)


https://github.com/duckdb/duckdb/pull/18135 introduced the `MERGE INTO`
operator.
Merge allows us to do upserts - similar to `ON CONFLICT`, but without
requiring the target table to have an index.

The merge operator has a separate implementation that leans on the
existing insert/update/delete operators.
The `ON CONFLICT` instead has a fully custom implementation that inlines
upserting into the `PhysicalInsert`.

This has a number of issues:
* The `ON CONFLICT` code is complex, and we now have an alternative code
path that can be used to achieve the same functionality. Maintaining
both is sub-optimal.
* The `ON CONFLICT` code always uses the index for detecting conflicts -
effectively forcing an index join to be used

This PR unifies the implementation of these two by rewriting the `ON
CONFLICT` clause in the binder to an identical `MERGE` statement, i.e.
this:

```sql
INSERT OR REPLACE INTO tbl FROM source;
```

Becomes (under the hood):

```sql
MERGE INTO tbl USING (SELECT DISTINCT ON (key) * FROM source) ON (tbl.key = source.key)
WHEN MATCHED THEN UPDATE
WHEN NOT MATCHED THEN INSERT
```

### Behavioral Changes

#### Multi-Row Updates

The previous upsert implementation threw an error if the same row was
updated multiple times in the same statement. This functionality has
been removed - this is now allowed.

#### WHERE clause changes

The ON CONFLICT clause in Postgres has the ability to specify a WHERE
clause after the indexed columns as part of the conflict target, e.g.:

```sql
ON CONFLICT (a) WHERE index_predicate [conflict_action]
```

This predicate is used to refer to partial indexes. 

> index_predicate
> Used to allow inference of partial unique indexes. Any indexes that
satisfy the predicate (which need not actually be partial indexes) can
be inferred. Follows CREATE INDEX format. SELECT privilege on any column
appearing within index_predicate is required.

DuckDB does not support partial indexes. However, this WHERE clause was
instead mistakenly used as a WHERE clause on the condition.

There is a second WHERE clause that can be specified as part of DO
UPDATE SET that is instead used for this:

```sql
    DO UPDATE SET ...
              [ WHERE condition ]
```
 
After this PR, we now throw an error if the WHERE clause is used as part
of the conflict target as we don't support partial indexes - and instead
only support WHERE as part of DO UPDATE SET.

#### WHERE fallback

In the previous implementation, if the WHERE clause was not met, we
would fallback to inserting. This is not particularly useful, as this
would always result in a constraint violation. After all: the `CONFLICT`
clause matched for that particular row. This also does not match
Postgres' behavior - they fall back to `DO NOTHING` (i.e. ignoring the
row). We now implement the same fallback as Postgres.

We also no longer support WHERE as part of `DO NOTHING`, as that does
not have any purpose anymore (the fallback would also be `DO NOTHING`).

### (De)serialization

As we do the rewrite in the binder, this does not apply when
deserializing plans from older versions. As such in order to support
deserializing older plans, for now the old code is still kept in. It
should just not be triggered anymore except when deserializing old
plans.

### Benchmarks

The `MERGE` code-path automatically switches between an index join
(using pushdown) and a full hash join - allowing it to be much faster
for bulk appends. Below is a benchmark running `INSERT OR REPLACE` on a
full TPC-H SF1 lineitem table with primary key constraint defined:

```sql
INSERT OR REPLACE INTO target_lineitem FROM lineitem
```

| main  | New  |
|-------|------|
| 23.2s | 2.3s |

tpot committed:
Change ctrl-a/ctrl-e to move to start/end of line, not buffer

samansmink committed:
add delta linux to ci

hannes committed:
nine is the magic number

Mytherin committed:
More insights around dict_fsst compression failure (#18475)

Mytherin committed:
Change ctrl-a/ctrl-e to move to start/end of line, not buffer (#18490)

This cli behaviour has been driving me crazy, but perhaps it's only me.
I think that for a multi-line query, you should go to the start/end of
the current line with ctrl-a/ctrl-e, not start/end of the entire
multi-line buffer. If you want to go to the start/end of the buffer, the
home/end keys are bound to that.

It could be a personal preference of mine, or maybe I spend too much
time in the cli, but I think this is more intuitive behaviour for
multi-line queries as it mirrors how multi-line editing works in some
other apps, e.g web browser, word processors etc.

hannes committed:
format

Mytherin committed:
CSVBuffer: check if block is loaded in CSVBuffer::Pin

Mytherin committed:
add delta linux back to ci (#18491)

This should work fine, it was disabled before due running out of disk
space.

Mytherin committed:
Prevent rolling back already committed transaction in the MetaTransaction

Mytherin committed:
Still set prev pointer in next segment

Mytherin committed:
Add missing template

Mytherin committed:
Use AddBlob in DictFSST - it can also be used to store blobs

Mytherin committed:
Align RLE count downwards for alignment purposes

Dtenwolde committed:
Remove assert and add test cases

aplavin committed:
bump minimal supported Julia version

aplavin committed:
restore previously existing test

Mytherin committed:
In ParseInfo::QualifierToString, also stringify schema when it is the default schema

Mytherin committed:
Only do parsed statement verification for query verification, clean up verify_fetch_row skipped tests

Mytherin committed:
Remove unnecessarily skipped tests

Mytherin committed:
Oss-Fuzz cases + clean-up

Mytherin committed:
Fix accidental internal exception in type transformation (#18492)

We had a bug in the error message construction if too many type
parameters were passed (#18452). This PR fixes this.

Mytherin committed:
Increase depth for correlated subqueries correctly, and keep depth + root binder stored in the binder itself

Mytherin committed:
Add new ossfuzz cases

Mytherin committed:
Merge branch 'main' into bugfixes

rustyconover committed:
fix: add missing space in AttachInfo::ToString()

taniabogatsch committed:
Merge branch 'v1.3-ossivalis' into merge-ossivalis

# Conflicts:
#	src/transaction/duck_transaction_manager.cpp

Mytherin committed:
Revert collate fix

Mytherin committed:
Format fix

Mytherin committed:
[Profiling] Add client context into read functions (#18438)

- Follow-up to https://github.com/duckdb/duckdb/pull/17754 and
https://github.com/duckdb/duckdb/pull/17875


Related issue:
https://github.com/duckdblabs/duckdb-internal/issues/4839.

Mytherin committed:
julia: config improvements (#17585)

Support these forms:
- `DuckDB.Config(option=value)` and `Config(["option" => "value"])`
instead of requiring separate `Config()` + set options one by on.
- `DuckDB.DB(..., config=(;option=value))`
- special `readonly` argument, `DuckDB.DB(..., readonly=true)`, that
already exists in the python client

Mytherin committed:
fix: add missing space in AttachInfo::ToString() (#18500)

Just improving the ToString() for AttachInfo to add a missing space.

Mytherin committed:
Merge ossivalis (#18502)

Mytherin committed:
Merge branch 'main' into bugfixes

Mytherin committed:
Change UNICODE to UTF8 (#17586)

Changed enum value UnicodeType::UNICODE to UnicodeType::UTF8 to avoid
conflict with /DUNICODE on MSVC

Mytherin committed:
Fix: Remove overly strict assertion on empty string value (#18504)

Fixes https://github.com/duckdblabs/duckdb-internal/issues/5475

This removes the assertion `D_ASSERT(!entry_value.empty())`.

This assertion was based on the assumption that an empty string was an
invalid state at this point. However, I think empty strings are a valid
case and are handled gracefully by downstream code. The assertion was
therefore too aggressive and has been removed.

Mytherin committed:
Fix several bugs/fuzzer issues (#18503)

* CSVBuffer: check if block is loaded in CSVBuffer::Pin
* Prevent rolling back already committed transaction in the
MetaTransaction
* Use AddBlob in DictFSST - it can also be used to store blobs which are
invalid UTF8
* Align RLE count downwards to ensure pointers are aligned
* Only do parsed statement verification for query verification, clean up
verify_fetch_row skipped tests
* Increase depth for correlated subqueries correctly, and keep depth +
root binder stored in the binder itself
* Re-organize oss-fuzz cases to be in sequential order instead of using
the random case numbers

VGSML committed:
feat: rebase

VGSML committed:
Merge remote-tracking branch 'origin/feat/table-udf-client-context' into feat/table-udf-client-context

# Conflicts:
#	src/include/duckdb.h
#	src/include/duckdb/main/capi/extension_api.hpp

taniabogatsch commented:
I think something must've gone wrong on the merge here?

VGSML commented:
Yes, I have made a new one PR
